### PR TITLE
Fix concurrency counting for windows systems with more than 1 processor group

### PR DESCRIPTION
**Pull request by Disservin**
fixes https://github.com/official-stockfish/Stockfish/issues/5307

### CODE REVIEW

**Comment by Sopel97** [comment]
are we sure std::thread::hardware_concurrency is returning an incorrect value?

**Comment by Disservin** [comment]
fairly certain according to so https://stackoverflow.com/questions/70704458/stdthreadhardware-concurrency-does-not-return-the-correct-number-of-logica

**Comment by Disservin** [comment]
https://github.com/microsoft/STL/discussions/1230

File: src/misc.cpp [hunk 1]
```diff
@@ -34,16 +34,9 @@
...
-using fun2_t = bool (*)(USHORT, PGROUP_AFFINITY);
-using fun3_t = bool (*)(HANDLE, CONST GROUP_AFFINITY*, PGROUP_AFFINITY);
-using fun4_t = bool (*)(USHORT, PGROUP_AFFINITY, USHORT, PUSHORT);
-using fun5_t = WORD (*)();
-using fun6_t = bool (*)(HANDLE, DWORD, PHANDLE);
-using fun7_t = bool (*)(LPCSTR, LPCSTR, PLUID);
-using fun8_t = bool (*)(HANDLE, BOOL, PTOKEN_PRIVILEGES, DWORD, PTOKEN_PRIVILEGES, PDWORD);
+using fun1_t = bool (*)(HANDLE, DWORD, PHANDLE);
+using fun2_t = bool (*)(LPCSTR, LPCSTR, PLUID);
+using fun3_t = bool (*)(HANDLE, BOOL, PTOKEN_PRIVILEGES, DWORD, PTOKEN_PRIVILEGES, PDWORD);
```
Comment (Sopel97) [inline src/misc.cpp:39]: if we're changing these unrelated things then it would probably be a good idea to actually name them

File: src/numa.h [hunk 2]
```diff
@@ -70,8 +74,54 @@ namespace Stockfish {
...
+    concurrency = 0;
+
+    DWORD offset = 0;
+
+    while (offset < length)
+    {
+        if (ptr->Relationship == RelationProcessorCore)
+            for (WORD group = 0; group < ptr->Processor.GroupCount; ++group)
+                for (KAFFINITY mask = ptr->Processor.GroupMask[group].Mask; mask != 0; mask >>= 1)
+                    concurrency += mask & 1;
```
Comment (Sopel97) [inline src/numa.h:112]: we can use popcount

**Review by Sopel97 — COMMENTED** [review]
Ok, tangentially related to this, we appear to have some more issues with respecting processor affinity on Windows than I thought. My understanding of how `GetNumaProcessorNodeEx` works was incorrect, and so I'll need to change things a bit in a later PR. I'm writing this because our currently half-broken implementation of handling processor affinity (taskset) on Windows relied on std::thread::hardware_concurrency returning the number of processors with respect to affinity set by taskset, which the new implementation omits, so we're going from a half-broken implementation to a differently half-broken implementation. I think it's ok, and another PR is required to fix the behaviour fully.

**Comment by Sopel97** [comment]
Looks like there might be a better way to implement this https://devblogs.microsoft.com/oldnewthing/20200824-00/?p=104116

**Comment by Sopel97** [comment]
so now I'm curious if std::thread::hardware_concurrency also respects affinity on other systems, which could potentially also break some stuff

**Comment by vondele** [comment]
as we go through the effort of calling lscpu, and respect taskset, probably safer to implement this based on that data (and the equivalent on windows) ?

**Comment by Sopel97** [comment]
yes, but on systems other than linux and windows we rely on std::thread::hardware_concurrency for processor numbers, though this issue is minor because we never bind threads on these systems.

**Comment by Disservin** [comment]
how do i fix this for win32?

**Comment by Sopel97** [comment]
I'm hitting some issue on the 32-bit build in my PR too. We might need to exclude 32-bit windows from this

**Comment by vondele** [comment]
I don't think we need numa support on 32 bit windows.

**Comment by Sopel97** [comment]
no support for processor groups even, which is most likely what's causing issues

**Comment by Sopel97** [comment]
vscode file in commit

**Comment by Disservin** [comment]
gr how that end up here

File: src/misc.cpp [hunk 3]
```diff
@@ -488,22 +482,25 @@ static void* aligned_large_pages_alloc_windows([[maybe_unused]] size_t allocSize
...
         return nullptr;
 
     // We need SeLockMemoryPrivilege, so try to enable it for the process
-    if (!fun6(  // OpenProcessToken()
+    if (!OpenProcessToken_f(  // OpenProcessToken()
           GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, &hProcessToken))
         return nullptr;
 
-    if (fun7(  // LookupPrivilegeValue(nullptr, SE_LOCK_MEMORY_NAME, &luid)
+    if (LookupPrivilegeValueA_f(  // LookupPrivilegeValue(nullptr, SE_LOCK_MEMORY_NAME, &luid)
```
Comment (Sopel97) [inline src/misc.cpp:503]: we don't need these comments now

**Pull request closed without merging**